### PR TITLE
CSV output, all variants (not just products), configurable FTP server and directory

### PR DESCRIPTION
**Pull request by martonpe**
This is all the changes that were needed to make this work for us.

### CODE REVIEW

File: lib/spree_google_base/feed_builder.rb [hunk 1]
```diff
@@ -1,22 +1,33 @@
...
-        exporter.generate_xml file
+      format = filename.split('.')[-1]
+      if format == "xml"
+        File.open(exporter.path, "w") do |file|
+          exporter.generate_xml file
+        end
+      elsif format == "txt"
+        File.open(exporter.path, "w") do |file|
+          file.write(exporter.generate_txt)
+        end
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: You should DRY this up, validate the format, open the file and call the proper method.

`generate_txt` should have the same signature as `generate_xml` and take an IO object. This would also allow you to define a method `generate(format, io)` and move the validation inside.

There should always be only one way to call methods that essentially do the same (`generate_xml` and `generate_txt`)

File: lib/spree_google_base/feed_builder.rb [hunk 2]
```diff
@@ -35,31 +46,36 @@ def initialize(opts = {})
...
+      delete_file_if_exists
 
-      File.open(path, 'w') do |file| 
-        generate_xml file
+      File.open(path, 'w') do |file|
+        if @format == 'xml'
+          generate_xml file
+        else
+          file.write(generate_txt)
+        end
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: See my comment above.

File: lib/spree_google_base/feed_builder.rb [hunk 3]
```diff
@@ -84,60 +100,106 @@ def generate_xml output
...
         end
       end
     end
-    
-    def transfer_xml
+
+    def generate_txt
+      csv_string = CSV.generate(col_sep: "\t") do |csv|
+        # Header row
+        additional_attrs = ["link", "image_link"]
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: You can use `%w[]` here which makes for a better reading.

File: lib/spree_google_base/feed_builder.rb [hunk 4]
```diff
@@ -84,60 +100,106 @@ def generate_xml output
...
+            build_variant_xml(xml, variant)
           end
         end
       end
     end
-    
-    def transfer_xml
+
+    def generate_txt
+      csv_string = CSV.generate(col_sep: "\t") do |csv|
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: You should wrap an IO object instead to conform to the same interface as `generate_xml`.

Check out http://ruby-doc.org/stdlib-2.1.0/libdoc/csv/rdoc/CSV.html#class-CSV-label-Wrap+an+IO+Object

File: lib/spree_google_base/feed_builder.rb [hunk 5]
```diff
@@ -84,60 +100,106 @@ def generate_xml output
...
-    def build_images(xml, product)
+
+    def build_variant_txt(variant)
+      attr_array = []
+      GOOGLE_BASE_ATTR_MAP.each do |k, v|
+        value = variant.send(v)
+        attr_array << value
+      end
+      attr_array << product_url(variant.slug, :host => domain)
+      attr_array.concat(build_images_txt(variant))
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: These two lines are *magic* because it's unclear what fields are coming. You can extract the line:

`additional_attrs = ["link", "image_link"]`

into a constant (`GOOGLE_BASE_ADDITIONAL_ATTRS`) and do a `case...when` inside a similar loop then above. This will allow the order of fields changed independently from the actual values. You don't want to have any future change that would require changing code in more than on place.

File: lib/spree_google_base/feed_builder.rb [hunk 6]
```diff
@@ -84,60 +100,106 @@ def generate_xml output
...
+      if Spree::GoogleBase::Config[:enable_additional_images]
+        more_images.each do |image|
+          xml.tag!('g:additional_image_link', image_url(variant, image))
+        end
+      end
+    end
 
+    def build_images_txt(variant)
+      images = []
       if Spree::GoogleBase::Config[:enable_additional_images]
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: You do not need this if statement, `main_image` will take the correct value irrespective of whether you want more images or not.

File: lib/tasks/spree_google_base_extension_tasks.rake [hunk 7]
```diff
@@ -1,13 +1,17 @@
 require 'net/ftp'
 
 namespace :spree_google_base do
-  task :generate_and_transfer => [:environment] do |t, args|
-    SpreeGoogleBase::FeedBuilder.generate_and_transfer
+  desc "Generate and transfer product feed in specified format. Formats: xml, tab separated txt"
```
Comment (rgabo) [inline lib/tasks/spree_google_base_extension_tasks.rake]: I think you can just say xml and txt here, no need to call out tab separated. Google calls the format txt themselves.

File: lib/tasks/spree_google_base_extension_tasks.rake [hunk 8]
```diff
@@ -1,13 +1,17 @@
...
+  task :generate_and_transfer, [:format] => [:environment] do |t, args|
+    raise 'Invalid format specified! Supported formats: xml, txt.' unless args[:format] == 'txt' or args[:format] == 'xml'
```
Comment (rgabo) [inline lib/tasks/spree_google_base_extension_tasks.rake]: `unless %w[txt xml].include? args[:format]`

File: lib/spree_google_base/feed_builder.rb (hunk 2 above)
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: You should definitely push `format` further down. You don't need to be aware of the format in this method.

**Comment by rgabo** [comment]
LGTM, awesome PR. The only concern I have is that `spree-google-base` now only supports variant-level feeds, but upstream supports product-level feeds.

There will be spree stores that do not use the variant level and this will be a breaking change for them. If we want to submit a PR upstream that we want to merge, we need to keep the product-level feed support. Let's prioritize that work.

File: lib/spree_google_base/feed_builder.rb (hunk 1 above)
Comment (martonpe) [inline lib/spree_google_base/feed_builder.rb]: Good point. I meant to clean this up, but then I got too happy that it was working :wink:

**Comment by martonpe** [comment]
@rgabo Had a look at the current feed:
* The feed does currently include a master variant for all products, even the ones that have other variants. This is not how it is supposed to work, so I'll fix this.
* All products have a master variant, so for users who don't have variants this will still export all the products. The only scenario where this could break things is if someone had variants in the store, but used this extension to only export the products and doesn't want the variants included.
* This IS a breaking change, though for two reasons: 
  * the rake task needs a parameter now. I could easily include a default, though. 
  * custom parameter definitions must go to variant_decorator now, and I don't think there is a way to get around that.

So we might not be able to submit this to upstream...

**Comment by martonpe** [comment]
@rgabo I decided to initialize the builders with the format parameter, because it's used throughout the class during generation, so this makes a lot more sense to me then passing it around a dozen times. :wink:

I'm not sure what you mean with the `GOOGLE_BASE_ADDITIONAL_ATTRS` stuff, let's discuss offline.

File: lib/spree_google_base/feed_builder.rb [hunk 9]
```diff
@@ -1,31 +1,33 @@
...
-    def self.generate_and_transfer
-      self.builders.each do |builder|
+
+    GOOGLE_BASE_ADDITIONAL_ATTRS = %w[link image_link]
+    GOOGLE_BASE_ADDITIONAL_ATTRS <<  "additional_image_link" if Spree::GoogleBase::Config[:enable_additional_images]
+
+    attr_reader :store, :domain, :title, :format
+
+    def self.generate_and_transfer(format)
+      raise "Invalid format specified! Supported formats: xml, txt" unless %w[txt xml].include? format
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: You should validate this in the constructor of `FeedBuilder` and only there so that no matter what code path, a `FeedBuilder` cannot be initialized with an invalid format.

File: lib/spree_google_base/feed_builder.rb [hunk 10]
```diff
@@ -68,76 +67,149 @@ def path
...
+    def generate_file
+      File.open(path, "w") do |file|
+        case
+        when @format == "xml"
+          generate_xml file
+          path
+        when @format == "txt"
+          generate_txt file
+          path
+        end
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: Let's use `case ... when` for what it was meant for:

```
case @format
when "xml" then generate_xml(file)
when "txt" then generate_txt(file)
end
path
```

I would also define an `attr_reader :format` so that it is easily accessible both from within and outside the object.

File: lib/spree_google_base/feed_builder.rb [hunk 11]
```diff
@@ -68,76 +66,146 @@ def path
...
+
+    def bulid_additional_images_txt(variant)
+      more_images = get_additional_images(variant)
+
+      return unless more_images
+      images = []
+      more_images.each do |image|
+        images << image_url(variant, image)
+      end
+      images.join(',')
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: More idiomatic Ruby:

```
if more_images = get_additional_images(variant)
  more_images.map { |image| image_url(variant, image) }.join(',')
end
```

File: lib/spree_google_base/feed_builder.rb [hunk 12]
```diff
@@ -68,76 +66,146 @@ def path
...
       end
     end
 
-    def image_url product, image
-      base_url = image.attachment.url(product.google_base_image_size)
+    def build_image_txt(variant)
+      main_image = get_main_image(variant)
+
+      return unless main_image
+      image_url(variant, main_image)
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: `return unless` is hard to follow, `if main_image = ...` will return `nil` if there is no main image and there is no `else` branch.

File: lib/spree_google_base/feed_builder.rb [hunk 13]
```diff
@@ -68,76 +66,146 @@ def path
...
+      GOOGLE_BASE_ATTR_MAP.each do |k, v|
+        value = variant.send(v)
+        attr_array << value
       end
+      GOOGLE_BASE_ADDITIONAL_ATTRS.each do |attribute|
+        case
+        when attribute == "link" then attr_array << product_url(variant.slug, :host => domain)
+        when attribute == "image_link" then attr_array << build_image_txt(variant)
+        when attribute == "additional_image_link" then attr_array << bulid_additional_images_txt(variant)
+        end
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: Sorry, but I need to point it out that you're not using `case .. when` properly.

`case` and `switch` in all languages is meant to branch based on the value of something. The current code is just a fancy way of writing `if` statements. Instead, you should:

```
case attribute
when "link" then
when "image_link" then
end
```

File: lib/spree_google_base/feed_builder.rb [hunk 14]
```diff
@@ -68,76 +66,146 @@ def path
...
       xml.item do
-        xml.tag!('link', product_url(product.slug, :host => domain))
-        build_images(xml, product)
-        
+        GOOGLE_BASE_ADDITIONAL_ATTRS.each do |attribute|
+          case
+          when attribute == "link" then xml.tag!('link', product_url(variant.slug, :host => domain))
+          when attribute == "image_link" then build_image_xml(xml, variant)
+          when attribute == "additional_image_link" then build_additional_images_xml(xml, variant)
+          end
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: Please see my comment below about `case ... when `statements`.

File: lib/spree_google_base/feed_builder.rb [hunk 15]
```diff
@@ -68,76 +66,146 @@ def path
...
-        main_image, *more_images = product.master.images
-      else
-        main_image = product.master.images.first
+
+    def build_variant_txt(variant)
+      attr_array = []
+      GOOGLE_BASE_ATTR_MAP.each do |k, v|
+        value = variant.send(v)
+        attr_array << value
       end
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]:
```
attributes = GOOGLE_BASE_ATTR_MAP.map { |_, v| variant.send(v) }
additional_attributes = GOOGLE_BASE_ADDITIONAL_ATTRS.map do |attribute|
  ...
end
attributes + additional_attributes
```

File: lib/spree_google_base/feed_builder.rb [hunk 16]
```diff
@@ -68,76 +66,146 @@ def path
...
+      GOOGLE_BASE_ADDITIONAL_ATTRS.each do |attribute|
+        case
+        when attribute == "link" then attr_array << product_url(variant.slug, :host => domain)
+        when attribute == "image_link" then attr_array << build_image_txt(variant)
+        when attribute == "additional_image_link" then attr_array << bulid_additional_images_txt(variant)
+        end
+      end
+      return attr_array
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: Explicit `return` is not needed here.

File: lib/spree_google_base/feed_builder.rb [hunk 17]
```diff
@@ -68,76 +66,146 @@ def path
...
       File.delete(path) if File.exists?(path)
     end
 
-    def generate_xml output
+    def generate_file
+      File.open(path, "w") do |file|
+        case
+        when @format == "xml" then generate_xml file
+        when @format == "txt" then generate_txt file
+        end
```
Comment (rgabo) [inline lib/spree_google_base/feed_builder.rb]: You can call the correct method dynamically here if you want:

`send("generate_#{@format}", file)`

and no need for branching.

Comment (martonpe) [inline lib/spree_google_base/feed_builder.rb]: I've corrected the case .. when here, but didn't use the dynamic method call magic. I don't think it's justified for something this simple.

**Comment by rgabo** [comment]
@martonpe merged.

`attr_reader :format` could make things easier on the eye and make the `format` available outside `FeedBuilder` but no strong preference.
